File: app/models/sql_workbench.py
from datetime import datetime, timezone

from sqlalchemy import Boolean, CheckConstraint, Column, DateTime, Integer, String, Text, ForeignKey
from .base import Base


class SqlDraft(Base):
    """查询草稿"""
    __tablename__ = "sql_draft"

    id = Column(Integer, primary_key=True, autoincrement=True)
    title = Column(String(200), nullable=False, default="", comment="草稿标题，空字符串表示未命名")
    sql_text = Column(Text, nullable=False, comment="SQL 文本")
    datasource_id = Column(Integer, ForeignKey("datasource_config.id"), nullable=True, comment="关联数据源")
    dialect = Column(String(50), nullable=False, default="oracle", comment="SQL 方言")
    description = Column(Text, nullable=True, comment="描述")
    tags = Column(Text, nullable=True, comment="标签（JSON array）")
    is_template = Column(Boolean, nullable=False, default=False, comment="是否模板")
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))


class SqlExecutionHistory(Base):
    """执行历史"""
    __tablename__ = "sql_execution_history"

    id = Column(Integer, primary_key=True, autoincrement=True)
    sql_text = Column(Text, nullable=False, comment="原始 SQL")
    sql_hash = Column(String(64), nullable=False, comment="SHA256(去首尾空白SQL)")
    datasource_id = Column(Integer, ForeignKey("datasource_config.id"), nullable=True, comment="数据源ID")
    datasource_name = Column(String(100), nullable=True, comment="执行时数据源名称（快照）")
    status = Column(String(20), nullable=False, default="success", comment="success / error / cancelled")
    elapsed_ms = Column(Integer, nullable=True, comment="执行耗时（毫秒）")
    row_count = Column(Integer, nullable=True, comment="返回行数")
    truncated = Column(Boolean, nullable=False, default=False, comment="是否因 LIMIT 截断")
    error_message = Column(Text, nullable=True, comment="错误信息")
    created_at = Column(DateTime, nullable=False, default=lambda: datetime.now(timezone.utc))

    __table_args__ = (
        CheckConstraint("status IN ('success', 'error', 'cancelled')", name="ck_history_status"),
    )

File: frontend/src/api/sqlWorkbench.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiFetch } from './client'

/* ─── Types ─── */

export interface DatasourceOption {
  id: number
  name: string
  ds_type: string
  dialect: string
}

export interface SchemaNode {
  schema_name: string
  tables: TableNode[]
}

export interface TableNode {
  id: number
  name: string
  comment: string | null
  column_count: number
}

export interface SchemaTree {
  datasource_id: number
  datasource_name: string
  schemas: SchemaNode[]
}

export interface ColumnDetail {
  id: number
  name: string
  type: string
  nullable: boolean
  comment: string | null
  is_primary_key: boolean
  is_foreign_key: boolean
}

export interface SearchResult {
  match_type: 'table' | 'column'
  schema_name: string
  table_name: string
  table_comment: string | null
  column_name: string | null
  table_id: number
}

export interface ExecuteInput {
  datasource_id: number
  sql: string
}

export interface ExecuteResult {
  columns: string[]
  rows: any[][]
  row_count: number
  truncated: boolean
  elapsed_ms: number
  error?: string | null
  history_id?: number | null
}

export interface SqlDraft {
  id: number
  title: string
  sql_text: string
  datasource_id: number | null
  dialect: string
  description: string | null
  tags: string | null
  is_template: boolean
  created_at: string
  updated_at: string
}

export interface CreateDraftInput {
  title?: string
  sql_text: string
  datasource_id?: number | null
  dialect?: string
  description?: string | null
  tags?: string | null
}

export interface UpdateDraftInput {
  title?: string
  sql_text?: string
  datasource_id?: number | null
  description?: string | null
  tags?: string | null
}

export interface ExecutionHistory {
  id: number
  sql_text: string
  sql_hash: string
  datasource_id: number | null
  datasource_name: string | null
  status: string
  elapsed_ms: number | null
  row_count: number | null
  truncated: boolean
  error_message: string | null
  created_at: string
}

/* ─── Fetchers ─── */

function fetchDatasources(): Promise<DatasourceOption[]> {
  return apiFetch('/sql/datasources')
}

function fetchSchemaTree(datasourceId: number): Promise<SchemaTree> {
  return apiFetch(`/sql/schema?datasource_id=${datasourceId}`)
}

function fetchTableColumns(tableId: number): Promise<ColumnDetail[]> {
  return apiFetch(`/sql/tables/${tableId}/columns`)
}

function searchSchema(datasourceId: number, q: string): Promise<SearchResult[]> {
  return apiFetch(`/sql/schema/search?datasource_id=${datasourceId}&q=${encodeURIComponent(q)}`)
}

function executeSql(input: ExecuteInput): Promise<ExecuteResult> {
  return apiFetch('/sql/execute', {
    method: 'POST',
    body: JSON.stringify(input),
  })
}

function fetchDrafts(): Promise<SqlDraft[]> {
  return apiFetch('/sql/drafts')
}

function createDraft(input: CreateDraftInput): Promise<SqlDraft> {
  return apiFetch('/sql/drafts', {
    method: 'POST',
    body: JSON.stringify(input),
  })
}

function updateDraft(id: number, input: UpdateDraftInput): Promise<SqlDraft> {
  return apiFetch(`/sql/drafts/${id}`, {
    method: 'PUT',
    body: JSON.stringify(input),
  })
}

function deleteDraft(id: number): Promise<void> {
  return apiFetch(`/sql/drafts/${id}`, { method: 'DELETE' })
}

function fetchHistory(datasourceId?: number, limit = 50): Promise<ExecutionHistory[]> {
  const params = new URLSearchParams()
  if (datasourceId) params.set('datasource_id', String(datasourceId))
  params.set('limit', String(limit))
  return apiFetch(`/sql/history?${params}`)
}

/* ─── Hooks ─── */

export function useSqlDatasources() {
  return useQuery({
    queryKey: ['sql', 'datasources'],
    queryFn: fetchDatasources,
  })
}

export function useSchemaTree(datasourceId: number | null) {
  return useQuery({
    queryKey: ['sql', 'schema', datasourceId],
    queryFn: () => fetchSchemaTree(datasourceId!),
    enabled: !!datasourceId,
  })
}

export function useTableColumns(tableId: number | null) {
  return useQuery({
    queryKey: ['sql', 'columns', tableId],
    queryFn: () => fetchTableColumns(tableId!),
    enabled: !!tableId,
  })
}

export function useSearchSchema(datasourceId: number | null, q: string) {
  return useQuery({
    queryKey: ['sql', 'search', datasourceId, q],
    queryFn: () => searchSchema(datasourceId!, q),
    enabled: !!datasourceId && q.length > 0,
  })
}

export function useExecuteSql() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: executeSql,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['sql', 'history'] })
    },
  })
}

export function useSqlDrafts() {
  return useQuery({
    queryKey: ['sql', 'drafts'],
    queryFn: fetchDrafts,
  })
}

export function useCreateDraft() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: createDraft,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['sql', 'drafts'] })
    },
  })
}

export function useUpdateDraft() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ id, ...input }: { id: number } & UpdateDraftInput) => updateDraft(id, input),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['sql', 'drafts'] })
    },
  })
}

export function useDeleteDraft() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: deleteDraft,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['sql', 'drafts'] })
    },
  })
}

export function useSqlHistory(datasourceId?: number | null) {
  return useQuery({
    queryKey: ['sql', 'history', datasourceId],
    queryFn: () => fetchHistory(datasourceId ?? undefined),
  })
}
